add Unixlistener

LuckyTurtleDev opened the pull request:
replace: #51 
fix: #49

Change config:
move `port` and `listen` to a single new `listener` .
If `listener` start with with the prefix `unix://`, the listener will be interpreted as unix listener.

LuckyTurtleDev committed:
start unix listener

LuckyTurtleDev committed:
remove unwrap

LuckyTurtleDev committed:
fmt

LuckyTurtleDev committed:
Add OS matrix strategy to Rust CI workflow

LuckyTurtleDev committed:
remove warnings

LuckyTurtleDev committed:
fix non unix

LuckyTurtleDev committed:
fix win

LuckyTurtleDev committed:
fix test

LuckyTurtleDev committed:
update error message

LuckyTurtleDev commented:
i am not sure if i want to delete random files at start. It would be probably be better to just delete the socket at the drop implementation.

msrd0 commented:
No, you should not delete the socket file if it exists. Some daemon orchestrators like systemd will actually create the socket themselves and expect you to use that socket. I believe they also have a mode where they listen on the socket, and if a request comes in, starts the service on demand. But I have never used that feature of systemd, so I wouldn't know for sure, but I believe I read that systemd sets an environment variable so you'd know to re-use that socket, but then again I am against systemd-specific solutions, so I think we should just re-use the existing socket whenever we can.

Also, it should be possible to set user/group/permissions on the unix socket file. Especially when the socket was not created by systemd, this comes in handy. But this can be done in a follow-up PR.

LuckyTurtleDev committed:
delete socket on drop

LuckyTurtleDev committed:
fmt

LuckyTurtleDev committed:
Change test job to run exclusively on Ubuntu

Simplified the test job to run only on Ubuntu.

LuckyTurtleDev committed:
add logging

msrd0 committed:
Do not return `anyhow::Result` from `main`

The way Rust handles this is basically equivalent to panic'ing
at the location the error is returned. The console output is a
stack trace pointing to the code inside the main method. A clean
error message with an exit code 1 return is way better.

msrd0 committed:
Adjust the other main method as well

Why are there two? No way, three?

msrd0 committed:
ze englisch was not ze yellow from ze ägg

LuckyTurtleDev committed:
run ci?

LuckyTurtleDev committed:
Merge branch 'main' into unixlistener

LuckyTurtleDev committed:
check for signal

LuckyTurtleDev committed:
fix: exit directly on empty config

LuckyTurtleDev committed:
remove tokio-utils

msrd0 commented on src/api.rs:
```diff
@@ -217,8 +217,48 @@ pub(crate) async fn init(
+		if let Some(listener) = config.listener.strip_prefix("unix://") {
+			#[cfg(not(unix))]
+			{
+				anyhow::bail!("unix sockets is only supported on unix systems");
```
Ähm no, but I guess this restriction comes from the web framework?

Also, `s/sockets is/sockets are/`

LuckyTurtleDev replied:
if you mean the `strip_prefix` yes, the framework expect it without the prefix.

msrd0 replied:
No, I mean the _unix sockets are only supported on unix systems_ part. [Windows has support for Unix sockets since Windows 10](https://devblogs.microsoft.com/commandline/af_unix-comes-to-windows/).

LuckyTurtleDev replied:
Oh did not know this. However the web frame work do not allow me to use unix socket on windows, so their is nothing we can do about it.

msrd0 commented on src/api.rs:
```diff
@@ -217,8 +217,48 @@ pub(crate) async fn init(
+						if let Err(err) = remove_file(&self.0).with_context(|| {
+							format!("failed to remove file {:?}", self.0)
+						}) {
+							eprintln!("{err:?}");
```
Why suddenly `eprintln!` instead of one of the `log` macros?

msrd0 commented on src/api.rs:
```diff
@@ -217,8 +217,48 @@ pub(crate) async fn init(
+				Server::new(UnixListener::bind(listener))
+					.run(server)
+					.await?;
+				drop(delete_file);
```
This drop is pointless, isn't it? Because the scope ends in the next line anyways, so it would be automatically dropped?

LuckyTurtleDev replied:
yes. just a hint for me that the variable have a use.

msrd0 commented on src/main.rs:
```diff
@@ -450,20 +451,24 @@ async fn async_main(config: Config) {
+		 res  = async {
+			signal::ctrl_c().await.context("failed to listen for signal")
+		} => {res}
+	}?;
```
I think this whole block would be more readable if we used the [`select`](https://docs.rs/futures-util/latest/futures_util/future/fn.select.html) function instead of the macro. Also, rustfmt obviously gave up on the block, so if you keep this, please at least format it manually.

LuckyTurtleDev replied:
i agree that this is not pretty. however i do not like 2 select function inside each other either. I have apply some manual formatting.

LuckyTurtleDev committed:
fix test

LuckyTurtleDev committed:
apply review

LuckyTurtleDev committed:
adjust default config
